src/lib/operators/join_index.cpp

```diff
@@ -0,0 +1,346 @@
+#include "join_index.hpp"
+
+#include <map>
+#include <memory>
+#include <numeric>
+#include <set>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "all_type_variant.hpp"
+#include "resolve_type.hpp"
+#include "storage/index/base_index.hpp"
+#include "storage/iterables/create_iterable_from_column.hpp"
+#include "type_comparison.hpp"
+#include "utils/assert.hpp"
+#include "utils/performance_warning.hpp"
+
+namespace opossum {
+
+/*
+ * This is an index join implementation. It expects to find an index on the right column.
+ * It currently uses the index for the following join modes: inner, left, right, outer, natural, self, semi, anti.
+ * For the remaining join types or if no index is found it falls back to a nested loop join.
+ */
+
+JoinIndex::JoinIndex(const std::shared_ptr<const AbstractOperator> left,
+                     const std::shared_ptr<const AbstractOperator> right, const JoinMode mode,
+                     const std::pair<ColumnID, ColumnID>& column_ids, const PredicateCondition predicate_condition)
+    : AbstractJoinOperator(left, right, mode, column_ids, predicate_condition) {
+  DebugAssert(mode != JoinMode::Cross, "Cross Join is not supported by index join.");
+}
+
+const std::string JoinIndex::name() const { return "JoinIndex"; }
+
+std::shared_ptr<AbstractOperator> JoinIndex::recreate(const std::vector<AllParameterVariant>& args) const {
+  return std::make_shared<JoinIndex>(_input_left->recreate(args), _input_right->recreate(args), _mode, _column_ids,
+                                     _predicate_condition);
+}
+
+std::shared_ptr<const Table> JoinIndex::_on_execute() {
+  _create_table_structure();
+
+  _perform_join();
+
+  return _output_table;
+}
+
+void JoinIndex::_create_table_structure() {
```

Comment:
Currently, it is only common with the nested loop join, since we used it for orientation. The hash join and sort merge join are using a different interface, requiring a bigger refactoring, which is probably out of scope of this PR?

Reply:
This will likely change with #736 anyway